From 911eb0f9ac66aba5ddfa2a6464cc8b174497f7e4 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Sat, 15 Oct 2022 09:30:15 +0300
Subject: [PATCH] [SW-0] gaudi3: modify recovery check for boot desc failure

Currently, if boot descriptior initialization fails due to DMA, size
or descriptor validation, ROM enters recovery only as part of
ppboot bringup sequence. This means, ROM tries to verify manifest
image (if image validation is enabled) and also set boot info
(if boot info set is enabled) even though boot descriptor
initialization is failed. This patch add changes to check boot
descriptor init failure as part of arcboot_rom_check_n_enter_recovery
function.

Change-Id: I3037f00293b98473307682b92071d27389d3111f
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 241be71..321ddfc 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -1112,8 +1112,11 @@
 	     EFUSE_SYSCFG_GPIO02_RECOVERY_CONTROL) == 0)
 		override = regmap_readl(mmGPIO0_DATA) & FORCE_RECOVERY_MASK;
 
-	/* Enter recovery if automated GPIO[FORCE_RECOVERY_MASK] is asserted */
-	if (override == 0)
+	/*
+	 * Enter recovery if automated GPIO[FORCE_RECOVERY_MASK] is asserted
+	 * or boot descriptor initialization for SPI manifest is failed
+	 */
+	if ((override == 0) || (!arcboot_rom_boot_desc_init_done(HL_FW_BOOT_PTR_SPI)))
 		execute_recovery();
 }
 
